onnxscript/rewriter/ort_fusions/cos_sin_cache.py

```diff
@@ -47,13 +43,66 @@ def __init__(
         # matched nodes as part of the rewrite-step. We apply a separate final
         # pass to remove unused nodes.
         super().__init__(name, remove_nodes=False)
-        self._max_pos_id = max_pos_id
+        # TODO: Determine what should be the default max_pos_id value
+        self._max_pos_id = 2048  # Set a default max_pos_id value
+        self._max_pos_id_from_model = False
         # map from inv_freq to (cos, sin) values for transformed graph
         self._inv_freq_cos_sin_cache: dict[ir.Value, tuple[ir.Value, ir.Value]] = {}
         self._reshape = reshape
         self._cast = cast
         self._const_freqs = const_freqs
 
+    def _set_max_position_ids_from_model(self, model: ir.Model):
+        """Extract max_position_ids value from the metadata of an ONNX model."""
+        # TODO: Determine what the correct metadata key is for max_position_ids
+        if model.metadata_props["max_position_id"] is not None:
+            self._max_pos_id = model.metadata_props["max_position_id"]  # type: ignore[assignment]
+            self._max_pos_id_from_model = True
+
+    def _compute_const_freqs(self, op, freqs):
+        """Compute cos/sin values when frequencies are constant."""
+        angles = freqs.const_value.numpy()
+        cos_value = np.cos(angles)
+        sin_value = np.sin(angles)
+        cos_2d = op.Constant(value=ir.tensor(cos_value))
+        sin_2d = op.Constant(value=ir.tensor(sin_value))
+        return cos_2d, sin_2d
+
+    def _compute_dynamic_freqs(self, op, inv_freq, position_ids, dtype):
+        """Compute cos/sin values dynamically based on inv_freq and position_ids."""
+        if self._max_pos_id_from_model:
+            # Use max_pos_id from the model metadata
+            max_pos_id = self._max_pos_id
+        elif position_ids.const_value is not None:
+            # Calculate max_pos_id from the position_ids tensor
+            max_pos_id = int(np.max(position_ids.const_value.numpy()))
+            self._max_pos_id = max_pos_id
+        else:
+            # Dynamically compute max_pos_id from position_ids using ONNX ops
+            inv_freq = op.Reshape(inv_freq, op.Constant(value_ints=[1, -1]))
+            max_pos_id = op.ReduceMax(position_ids)
+            max_pos_id = op.Squeeze(max_pos_id, op.Constant(value_ints=[0]))
+            max_pos_id = op.Add(max_pos_id, op.Constant(value_int=1))
```

Comment:
What's the type of position_ids? Is it guaranteed to be int64?

Reply:
Why is there a +1 here, but not in the case down below where the value is known?

Reply:
Added a + 1 to the case down below